pkg/vm/stackitem/json.go

```diff
@@ -0,0 +1,202 @@
+package stackitem
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	gio "io"
+	"math"
+	"math/big"
+
+	"github.com/nspcc-dev/neo-go/pkg/io"
+)
+
+// decoder is a wrapper around json.Decoder helping to mimic C# json decoder behaviour.
+type decoder struct {
+	json.Decoder
+
+	depth int
+}
+
+// MaxAllowedInteger is the maximum integer allowed to be encoded.
+const MaxAllowedInteger = 2<<53 - 1
+
+// maxJSONDepth is a maximum allowed depth-level of decoded JSON.
+const maxJSONDepth = 10
+
+// ToJSON encodes Item to JSON.
+// It behaves as following:
+//   ByteArray -> base64 string
+//   BigInteger -> number
+//   Bool -> bool
+//   Null -> null
+//   Array -> array
+//   Map -> map with keys as UTF-8 bytes
+func ToJSON(item Item) ([]byte, error) {
+	buf := io.NewBufBinWriter()
+	toJSON(buf, item)
+	if buf.Err != nil {
+		return nil, buf.Err
+	}
+	buf.Len()
+	return buf.Bytes(), nil
+}
+
+func toJSON(buf *io.BufBinWriter, item Item) {
+	w := buf.BinWriter
+	if w.Err != nil {
+		return
+	} else if buf.Len() > MaxSize {
+		w.Err = errors.New("item is too big")
+	}
+	switch it := item.(type) {
+	case *Array, *Struct:
+		w.WriteB('[')
+		items := it.Value().([]Item)
+		for i, v := range items {
+			toJSON(buf, v)
+			if i < len(items)-1 {
+				w.WriteB(',')
+			}
+		}
+		w.WriteB(']')
+	case *Map:
+		w.WriteB('{')
+		for i := range it.value {
+			bs, _ := it.value[i].Key.TryBytes() // map key can always be converted to []byte
```

Comment:
Speaking of `Serialize` and `SerializeToByteArray` differences, this one probably is not true then.

Reply:
Why not? Map keys always have one of the primitive types.

Reply:
OK, they can, but they shouldn't be to match C# node behavior.

Reply:
https://github.com/neo-project/neo/issues/1704